File: clsTransformResults.cpp
// Written by Navdeep Jaitly and Anoop Mayampurath for the Department of Energy (PNNL, Richland, WA)
// Copyright 2006, Battelle Memorial Institute
// E-mail: navdeep.jaitly@pnl.gov
// Website: http://ncrr.pnl.gov/software
// -------------------------------------------------------------------------------
// 
// Licensed under the Apache License, Version 2.0; you may not use this file except
// in compliance with the License.  You may obtain a copy of the License at 
// http://www.apache.org/licenses/LICENSE-2.0

#include "clstransformresults.h"
#using <mscorlib.dll>
#include <algorithm>
namespace DeconToolsV2 
{
	namespace Results
	{
		clsTransformResults::clsTransformResults(void)
		{
			mobj_lcms_results = NULL ; 
			menmFileType = DeconToolsV2::Readers::FileType::UNDEFINED ; 
		}

		clsTransformResults::~clsTransformResults(void)
		{
			if (mobj_lcms_results != NULL)
			{
				delete mobj_lcms_results ; 
				mobj_lcms_results = NULL ; 
			}
		}
		
		void clsTransformResults::SetLCMSTransformResults(Engine::Results::LCMSTransformResults *results)
		{
			if (mobj_lcms_results != NULL)
			{
				delete mobj_lcms_results ; 
				mobj_lcms_results = NULL ; 
			}
			mobj_lcms_results = results ; 
		}

		int clsTransformResults::GetMinScan()
		{
			if(mobj_lcms_results == NULL)
				return -1 ; 
			return mobj_lcms_results->GetMinScan() ;
		}

		int clsTransformResults::GetMaxScan()
		{
			if(mobj_lcms_results == NULL)
				return -1 ; 
			return mobj_lcms_results->GetMaxScan() ;
		}

		bool clsTransformResults::IsDeisotoped()
		{
			if (mobj_lcms_results == NULL)
				throw gcnew System::Exception("No results stored.") ;
			return mobj_lcms_results->IsDeisotoped() ; 
		}

		
		// Used by the stl algorithm sort to sort std::vector of peaks in descending order of mdbl_intensity.
		bool RawPeaksIntensityComparison(Engine::Results::LCMSPeak<Engine::Results::PeakMinInfo> &pk1, Engine::Results::LCMSPeak<Engine::Results::PeakMinInfo> &pk2)
		{
			if (pk1.mobj_peak.mflt_intensity < pk2.mobj_peak.mflt_intensity)
				return true ; 
			if (pk1.mobj_peak.mflt_intensity > pk2.mobj_peak.mflt_intensity)
				return false ; 
			return pk1.mobj_peak.mflt_mz < pk2.mobj_peak.mflt_mz ; 
		}

		void clsTransformResults::GetRawDataSortedInIntensity(array<DeconToolsV2::Results::clsLCMSPeak^>^ (&lcms_peaks))
		{
			if (mobj_lcms_results == NULL)
			{
				lcms_peaks = nullptr ; 
				return ; 
			}

			std::vector<Engine::Results::LCMSPeak<Engine::Results::PeakMinInfo> > vectPeaks ; 
			int num_peaks = mobj_lcms_results->GetNumPeaks() ; 
			vectPeaks.reserve(num_peaks) ; 
			mobj_lcms_results->GetAllPeaks(vectPeaks) ; 
			std::sort(vectPeaks.begin(), vectPeaks.end(), &RawPeaksIntensityComparison) ; 

			lcms_peaks = gcnew array<DeconToolsV2::Results::clsLCMSPeak^>(num_peaks); 
			int min_scan = mobj_lcms_results->GetMinScan() ; 
			int max_scan = mobj_lcms_results->GetMaxScan() ; 
			Engine::Results::LCMSPeak<Engine::Results::PeakMinInfo> pk ; 

			for (int pk_num = 0 ; pk_num < num_peaks ; pk_num++)
			{
				mint_percent_done = (pk_num * 100)/num_peaks ; 
				pk = vectPeaks[pk_num] ; 
				lcms_peaks[pk_num] = gcnew clsLCMSPeak(pk.mint_scan_num, pk.mobj_peak.mflt_mz, pk.mobj_peak.mflt_intensity) ; 
			}
			mint_percent_done = 100 ; 
		}

		void clsTransformResults::GetRawData(array<DeconToolsV2::Results::clsLCMSPeak^>^ (&lcms_peaks))
		{
			if (mobj_lcms_results == NULL)
			{
				lcms_peaks = nullptr ; 
				return ; 
			}

			int num_peaks = mobj_lcms_results->GetNumPeaks() ; 
            lcms_peaks = gcnew array<DeconToolsV2::Results::clsLCMSPeak^>(num_peaks);
			int min_scan = mobj_lcms_results->GetMinScan() ; 
			int max_scan = mobj_lcms_results->GetMaxScan() ; 
			Engine::Results::LCMSPeak<Engine::Results::PeakMinInfo> pk ; 

			for (int pk_num = 0 ; pk_num < num_peaks ; pk_num++)
			{
				mint_percent_done = (pk_num * 100)/num_peaks ; 
				pk = mobj_lcms_results->GetPeak(pk_num) ; 
				lcms_peaks[pk_num] = gcnew clsLCMSPeak(pk.mint_scan_num, pk.mobj_peak.mflt_mz, pk.mobj_peak.mflt_intensity) ; 
			}
			mint_percent_done = 100 ; 
		}
		
		
        void clsTransformResults::GetSIC(int min_scan, int max_scan, float mz, float mz_tolerance, array<float> ^ (&peak_intensities))
		{
			std::vector<float> vect_intensities ; 
			mobj_lcms_results->GetSIC(min_scan, max_scan, mz-mz_tolerance, mz+mz_tolerance, vect_intensities) ; 
			int num_scans = max_scan - min_scan + 1 ; 
            peak_intensities = gcnew array<float>(num_scans);

			for (int scan_num = min_scan ; scan_num <= max_scan ; scan_num++)
			{
				peak_intensities[scan_num-min_scan] = vect_intensities[scan_num-min_scan] ; 
			}
		}

        void clsTransformResults::GetScanPeaks(int scan_num, array<float> ^ (&peak_mzs), array<float> ^ (&peak_intensities))
		{
			std::vector<float> vect_mzs ; 
			std::vector<float> vect_intensities ; 

			mobj_lcms_results->GetScanPeaks(scan_num, vect_mzs, vect_intensities) ; 

			int num_pts = vect_intensities.size() ; 
            peak_mzs = gcnew array<float>(num_pts);
            peak_intensities = gcnew array<float>(num_pts);

			for (int pt_num = 0 ; pt_num < num_pts ; pt_num++)
			{
				peak_mzs[pt_num] = vect_mzs[pt_num] ; 
				peak_intensities[pt_num] = vect_intensities[pt_num] ; 
			}
		}

		int clsTransformResults::GetNumPeaks()
		{
			return mobj_lcms_results->GetNumPeaks() ; 
		}

		void clsTransformResults::ReadResults(System::String ^fileName)
		{
			if (mobj_lcms_results == NULL)
			{
				mobj_lcms_results = new Engine::Results::LCMSTransformResults() ; 
			}

			char fileNameCh[512] ; 
			DeconEngine::Utils::GetStr(fileName, fileNameCh) ; 
			mobj_lcms_results->LoadResults(fileNameCh) ; 
		}

		void clsTransformResults::WriteResults(System::String ^fileName, bool save_signal_range)
		{
			char fileNameCh[512] ; 
			try
			{
				DeconEngine::Utils::GetStr(fileName, fileNameCh) ; 
				mobj_lcms_results->SaveResults(fileNameCh, save_signal_range) ; 
			}
			catch (char *mesg)
			{
				System::String ^exception_msg = gcnew System::String(mesg) ; 
				throw gcnew System::Exception(exception_msg) ; 
			}
		}

		void clsTransformResults::WriteScanResults(System::String ^fileName)
		{
			char fileNameCh[512] ; 
			try
			{
				DeconEngine::Utils::GetStr(fileName, fileNameCh) ; 
				if (menmFileType != DeconToolsV2::Readers::FileType::ICR2LSRAWDATA)
					mobj_lcms_results->SaveScanResults(fileNameCh, false) ; 
				else
					mobj_lcms_results->SaveScanResults(fileNameCh, true) ; 
			}
			catch (char *mesg)
			{
				System::String ^exception_msg = gcnew System::String(mesg) ; 
				throw gcnew System::Exception(exception_msg) ; 
			}
		}

	}
}

File: C#_Version/DeconEngine/clsTransformResults.cs
using System;
using System.Collections.Generic;
using Engine.Results;

namespace DeconToolsV2.Results
{

    /// <summary>
    /// Transform results
    /// </summary>
    /// <remarks>Used by Decon2LS.UI and by PeakImporter (from peak.dat files)</remarks>
    public class clsTransformResults
    {
        private Readers.FileType menmFileType = Readers.FileType.UNDEFINED;
        internal LCMSTransformResults mobj_lcms_results;


        public string FileName => mobj_lcms_results.GetFileName();

        public Readers.FileType FileType
        {
            get => menmFileType;
            set => menmFileType = value;
        }

        public float[] MonoMasses
        {
            get
            {
                var numDeisotoped = mobj_lcms_results.GetNumTransforms();
                var mono_masses = new float[numDeisotoped];
                for (var index = 0; index < numDeisotoped; index++)
                {
                    var fitRecord = mobj_lcms_results.GetIsoPattern(index);
                    mono_masses[index] = (float) fitRecord.MonoMw;
                }
                return mono_masses;
            }
        }

        public short[] Charges
        {
            get
            {
                var numDeisotoped = mobj_lcms_results.GetNumTransforms();
                var charges = new short[numDeisotoped];
                for (var index = 0; index < numDeisotoped; index++)
                {
                    var fitRecord = mobj_lcms_results.GetIsoPattern(index);
                    charges[index] = (short)fitRecord.ChargeState;
                }
                return charges;
            }
        }

        public float[] Scans
        {
            get
            {
                var numDeisotoped = mobj_lcms_results.GetNumTransforms();
                var scans = new float[numDeisotoped];
                for (var index = 0; index < numDeisotoped; index++)
                {
                    var fitRecord = mobj_lcms_results.GetIsoPattern(index);
                    scans[index] = fitRecord.ScanNum;
                }
                return scans;
            }
        }

        public clsTransformResults()
        {
            mobj_lcms_results = null;
        }

        internal void SetLCMSTransformResults(LCMSTransformResults results)
        {
            mobj_lcms_results = results;
        }

        public int GetMinScan()
        {
            if (mobj_lcms_results == null)
                return -1;
            return mobj_lcms_results.GetMinScan();
        }

        public int GetMaxScan()
        {
            if (mobj_lcms_results == null)
                return -1;
            return mobj_lcms_results.GetMaxScan();
        }

        public bool IsDeisotoped()
        {
            if (mobj_lcms_results == null)
                throw new Exception("No results stored.");
            return mobj_lcms_results.IsDeisotoped();
        }

#if Enable_Obsolete
        [Obsolete("Use GetRawDataSortedInIntensity that works on Engine.Results.LcmsPeak")]
        public void GetRawDataSortedInIntensity(out clsLCMSPeakOld[] lcms_peaks)
        {
            LcmsPeak[] lcmsPeaks;

            GetRawDataSortedInIntensity(out lcmsPeaks);

            lcms_peaks = ConvertLcmsPeaksToLCMSPeaks(lcmsPeaks);
        }
#endif

        public void GetRawDataSortedInIntensity(out LcmsPeak[] lcmsPeaks)
        {
            if (mobj_lcms_results == null)
            {
                lcmsPeaks = null;
                return;
            }

            var vectPeaks = new List<LcmsPeak>();
            var num_peaks = mobj_lcms_results.GetNumPeaks();
            vectPeaks.Capacity = num_peaks;
            mobj_lcms_results.GetAllPeaks(out vectPeaks);

            /*bool RawPeaksIntensityComparison(Engine.Results.LCMSPeak pk1,
                Engine.Results.LCMSPeak pk2)
            {
                if (pk1.mdbl_intensity < pk2.mdbl_intensity)
                    return true;
                if (pk1.mdbl_intensity > pk2.mdbl_intensity)
                    return false;
                return pk1.mdbl_mz < pk2.mdbl_mz;
            }*/
            // Used by the sort algorithm to sort List of peaks in descending order of mdbl_intensity.
            vectPeaks.Sort((x, y) =>
            {
                var result = y.Intensity.CompareTo(x.Intensity);
                if (result == 0)
                {
                    result = y.Mz.CompareTo(x.Mz);
                }
                return result;
            });

            lcmsPeaks = new LcmsPeak[num_peaks];

            for (var pk_num = 0; pk_num < num_peaks; pk_num++)
            {
                var pk = vectPeaks[pk_num];
                lcmsPeaks[pk_num] = new LcmsPeak(pk.ScanNum, pk.Mz, pk.Intensity);
            }
        }

#if Enable_Obsolete
        [Obsolete("Use GetRawDataSortedInIntensity that works on Engine.Results.LcmsPeak")]
        public void GetRawData(out clsLCMSPeakOld[] lcms_peaks)
        {
            LcmsPeak[] lcmsPeaks;

            GetRawDataSortedInIntensity(out lcmsPeaks);

            lcms_peaks = ConvertLcmsPeaksToLCMSPeaks(lcmsPeaks);
        }
#endif

        public void GetRawData(out LcmsPeak[] lcmsPeaks)
        {
            if (mobj_lcms_results == null)
            {
                lcmsPeaks = null;
                return;
            }

            var num_peaks = mobj_lcms_results.GetNumPeaks();
            lcmsPeaks = new LcmsPeak[num_peaks];

            for (var pk_num = 0; pk_num < num_peaks; pk_num++)
            {
                var pk = mobj_lcms_results.GetPeak(pk_num);
                lcmsPeaks[pk_num] = new LcmsPeak(pk.ScanNum, pk.Mz, pk.Intensity);
            }
        }

        public void GetSIC(int min_scan, int max_scan, float mz, float mz_tolerance, out float[] peak_intensities)
        {
            List<double> vect_intensities;
            mobj_lcms_results.GetSIC(min_scan, max_scan, mz - mz_tolerance, mz + mz_tolerance, out vect_intensities);
            var num_scans = max_scan - min_scan + 1;
            peak_intensities = new float[num_scans];

            for (var scan_num = min_scan; scan_num <= max_scan; scan_num++)
            {
                peak_intensities[scan_num - min_scan] = (float)vect_intensities[scan_num - min_scan];
            }
        }

        public void GetScanPeaks(int scan_num, out float[] peak_mzs, out float[] peak_intensities)
        {
            List<double> vect_mzs;
            List<double> vect_intensities;

            mobj_lcms_results.GetScanPeaks(scan_num, out vect_mzs, out vect_intensities);

            var num_pts = vect_intensities.Count;
            peak_mzs = new float[num_pts];
            peak_intensities = new float[num_pts];

            for (var pt_num = 0; pt_num < num_pts; pt_num++)
            {
                peak_mzs[pt_num] = (float)vect_mzs[pt_num];
                peak_intensities[pt_num] = (float)vect_intensities[pt_num];
            }
        }

        public int GetNumPeaks()
        {
            return mobj_lcms_results.GetNumPeaks();
        }

        [Obsolete("Only used by Decon2LS.UI (maybe), and by the PeakImporter (from peak.dat files)", false)]
        public void ReadResults(string fileName)
        {
            if (mobj_lcms_results == null)
            {
                mobj_lcms_results = new LCMSTransformResults();
            }

            mobj_lcms_results.LoadResults(fileName);
        }

        [Obsolete("Only used by Decon2LS.UI (maybe), and by the PeakImporter (from peak.dat files)", false)]
        public void WriteResults(string fileName, bool save_signal_range)
        {
            mobj_lcms_results.SaveResults(fileName, save_signal_range);
        }

        public void WriteScanResults(string fileName)
        {
#pragma warning disable 618
            if (menmFileType != Readers.FileType.ICR2LSRAWDATA)
#pragma warning restore 618
                mobj_lcms_results.SaveScanResults(fileName, false);
            else
                mobj_lcms_results.SaveScanResults(fileName, true);
        }

#if Enable_Obsolete
        [Obsolete("Conversion class; only used by obsolete methods")]
        private clsLCMSPeakOld[] ConvertLcmsPeaksToLCMSPeaks(IReadOnlyList<LcmsPeak> lcmsPeaks)
        {
            var lcms_peaks = new clsLCMSPeakOld[lcmsPeaks.Count];

            for (var i = 0; i < lcmsPeaks.Count; i++)
            {
                lcms_peaks[i] = new clsLCMSPeakOld(lcmsPeaks[i].ScanNum, lcmsPeaks[i].Mz, lcmsPeaks[i].Intensity);
            }
            return lcms_peaks;
        }
#endif

    }
}
